xapian-core/api/queryinternal.cc

```diff
@@ -865,94 +867,138 @@ Query::Internal::unserialise(const char ** p, const char * end,
+		    //   0000tttt where:
+		    //     tttt -> encodes which OP_XXX
+		    switch (ch & 0x0f) {
+			case 0x00: // OP_INVALID
+			    return new Xapian::Internal::QueryInvalid();
+			case 0x0a: { // Edit distance
+			    Xapian::termcount max_expansion;
+			    if (!unpack_uint(p, end, &max_expansion) ||
+						     end - *p < 2) {
+			    throw SerialisationError("not enough data");
+			    }
+			    int flags = static_cast<unsigned char>(*(*p)++);
+			    op combiner = static_cast<op>(*(*p)++);
+			    unsigned edit_distance;
+			    size_t fixed_prefix_len;
+			    string pattern;
+			    if (!unpack_uint(p, end, &edit_distance) ||
+				!unpack_uint(p, end, &fixed_prefix_len) ||
+				!unpack_string(p, end, pattern)) {
+				throw SerialisationError("not enough data");
+			    }
+			    using Xapian::Internal::QueryEditDistance;
+			    return new QueryEditDistance(pattern,
+							 max_expansion,
+							 flags,
+							 combiner,
+							 edit_distance,
+							 fixed_prefix_len);
+			}
+			case 0x0b: { // Wildcard
+			    Xapian::termcount max_expansion;
+			    if (!unpack_uint(p, end, &max_expansion) ||
+				end - *p < 2) {
+				throw SerialisationError("not enough data");
+			    }
+			    int flags = static_cast<unsigned char>(*(*p)++);
+			    op combiner = static_cast<op>(*(*p)++);
+			    string pattern;
+			    if (!unpack_string(p, end, pattern)) {
+				throw SerialisationError("not enough data");
+			    }
+			    return new Xapian::Internal::QueryWildcard(
+					    pattern,
+					    max_expansion,
+					    flags,
+					    combiner);
+			}
+			case 0x0c: { // PostingSource
+			    string name;
+			    if (!unpack_string(p, end, name)) {
+				throw SerialisationError("not enough data");
+			    }
+
+			    const PostingSource* reg_source =
+					    reg.get_posting_source(name);
+			    if (!reg_source) {
+				string m = "PostingSource ";
+				m += name;
+				m += " not registered";
+				throw SerialisationError(m);
+			    }
+
+			    string serialised_source;
+			    if (!unpack_string(p, end, serialised_source)) {
+				throw SerialisationError("not enough data");
+			    }
+			    PostingSource* source =
+				reg_source->unserialise_with_registry(
+						    serialised_source,
+						    reg);
+			    return new Xapian::Internal::QueryPostingSource
+						    (source->release());
+			}
+			case 0x0d: {
+			    using Xapian::Internal::QueryScaleWeight;
+			    double scale_factor = unserialise_double(p, end);
+			    return new QueryScaleWeight(scale_factor,
+				Query(unserialise(p, end, reg)));
+			}
+			case 0x0e: {
+			    Xapian::termcount wqf;
+			    Xapian::termpos pos;
+			    if (!unpack_uint(p, end, &wqf) ||
+				!unpack_uint(p, end, &pos)) {
+				throw SerialisationError("not enough data");
+			    }
+			    return new Xapian::Internal::QueryTerm(
+						    string(), wqf, pos);
+			}
+			case 0x0f:
+			    return new Xapian::Internal::QueryTerm();
+			default: // Others currently unused.
+			    break;
+			    }
```

Comment:
This still looks wrong. The close brace here is presumably closing the switch, so should be eight spaces left?